### PR TITLE
Add Missing URL state to Navigation Link Block

### LINKED ISSUE #23326: Navigation Screen: make it clear links need to be added for menu to be functional
**Is your feature request related to a problem? Please describe.**

Right now, when you're using the new navigation screen, nothing is stopping you from adding essentially empty links. Thinking like a new user, it's not necessarily clear that I would need to go back and link to things in the menu for it to truly work. In comparison, the previous menu screen had a built in flow that wouldn't allow a user to add an item to the menu without it first existing. 

Here's a 1 minute long video where I walk through both screens to better demonstrate the issue: https://cloudup.com/cXDDOWFgJO4

**Describe the solution you'd like**

I'd like to see prompts in the UI that ensure when someone adds an item to the menu, it's clear that a link needs to be added for it to be functional. Right now, the current flow makes it almost too easy to create empty menus. 

**Describe alternatives you've considered**

Not quite sure :)

### LINKED ISSUE #28440: Navigation: Reconsider placeholder state for the Link block
When you add a new Link block within the Navigation block, the Link UI appears. However, if you click anywhere or generally move focus away, the Link UI disappears and its difficult to determine how to complete setting up the Link block. This leads to empty Link blocks with placeholder "Add link" text that doesn't do anything.

https://user-images.githubusercontent.com/191598/105555316-21386480-5cd7-11eb-91e3-69d45a9fd549.mp4

I think it could be helpful to improve the Link block's placeholder state. For example, if you unselect a newly added Link block we could show a button to reopen the Link UI and add a URL:

https://user-images.githubusercontent.com/191598/105555509-88561900-5cd7-11eb-97ca-7ea964d59f8c.mp4

The experience for adding a Page or Post link could be adapted as well:

https://user-images.githubusercontent.com/191598/105555571-ab80c880-5cd7-11eb-9398-d2eb3073caa4.mp4

### PR DESCRIPTION
**Pull request by georgeh**
## Description
Changing the Navigation Link block to highlight when a link is missing 

Closes #28440
Closes #23326

## Screenshots 

https://user-images.githubusercontent.com/128240/108095707-28406180-704e-11eb-8d2a-981b6790a55c.mov

## Types of changes

When the URL is missing from a Navigation Link Block:
* Adds a placeholder
* Forces the URL popover open when the block is selected
* Prevents editing the text

Bugfix:
* Cancels any outstanding requests in `URLInput` when the component is unmounted to avoid calling `setState` on an unmounted component

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [na] ~~I've included developer documentation if appropriate.~~ 
- [na] ~~I've updated all React Native files affected by any refactorings/renamings in this PR.~~

### CODE REVIEW

**Comment by georgeh** [comment]
@jasmussen @shaunandrews This does a little bit with highlighting a Link block that is missing the URL and popping up the input fields. A really rough sketch, but please let me know if it's on the right track. (Setting the URL and text may or may not work)

**Comment by jasmussen** [comment]
Excited about this one! 

There's not a ton to see yet:

Screenshot 2021-02-09 at 12 16 08

One thing that I think is pretty key for the placeholder menu item, perhaps beyond opening the linkcontrol by default, is that the menu item label is not editable at all until both URL and Label have been filled out (or for Page link, a page has been chosen). That means the menu item is completely inert until configured. I suppose we can keep the RichText component and inert it using `pointer-events: none;` and `cursor: pointer;`, however that might not work as you could probably still tab into it. 

Can we remove the rich text part of the menu item, until configured?

File: packages/block-library/src/navigation-link/edit.js
Comment (shaunandrews) [inline packages/block-library/src/navigation-link/edit.js:381]: FWIW, I don't think we should stop using LinkControl, or necessarily change LinkControl to resolve #28440. The goal of that issue is really just about the interactions around a Navigation Link block when it doesn't have a link set; Showing the LinkControl when and "empty" Navigation Link is selected.

**Comment by jasmussen** [comment]
To Shaun's point, it could make sense to rewind this and make it even simpler:

- Insert menu item and the link control is immediately open with focus there. 
- You can't focus the menu item label, focus is always immediately transferred to the link control.
- Ideally, the menu item is inert until a URL has been added.
- Once a URL has been added, _now_ you can edit the label in the menu item, directly.

Did I get that right?

**Comment by shaunandrews** [comment]
Yes, exactly.

**Comment by georgeh** [comment]
Thanks! That's was the direction I've been heading in to see what we can do without overhauling too much, but that list works even better than where I was.

**Comment by georgeh** [comment]
OK, is this a bit closer to the interaction that we are trying to hit?

https://user-images.githubusercontent.com/128240/107568717-64d70d80-6bb5-11eb-8b2b-91feaaafd166.mov

**Comment by shaunandrews** [comment]
That seems to be on the right path! You can do anything until you select a link. Very nice.

**Comment by jasmussen** [comment]
Really nice work. I think the next step is to polish the visuals a bit. Shaun did you want to try that dimmed gray background from your mockup? Happy to give that a stab and push CSS here. 

Question: from looking at your video, the linkcontrol popover appears top/left aligned from where you click, whereas I believe it's top/center aligned in the master branch. Is this intentional and/or easy to change?

Bonus: can you make the linkcontrol use the `isAlternate` visual treatment? 

```
const POPOVER_PROPS = {
	isAlternate: true,
};
```

That would make it have a black border like the other block UI.

**Comment by georgeh** [comment]
> Question: from looking at your video, the linkcontrol popover appears top/left aligned from where you click, whereas I believe it's top/center aligned in the master branch. Is this intentional and/or easy to change?

I think this is a visual artifact because I was using a small viewport to record the video. I just tested it fullscreen and I think it's aligning correctly:

Screen Shot 2021-02-11 at 11 10 18 AM

If not, I'm not sure which part needs to move, maybe take a screenshot of the video? (That's also with `isAlternate` on the popover)

**Comment by georgeh** [comment]
@jasmussen do you want to take a stab at the visuals for the "Missing URL" state?

**Comment by jasmussen** [comment]
I pushed a little polish:

Screenshot 2021-02-15 at 11 09 10

It now has the default font and size, and a gray background reminiscent of the initial setup state.

@shaunandrews will be asking about the spacing here:

Screenshot 2021-02-15 at 11 09 03

That's the default spacing applied to `wp-block-navigation-link__content`. I the markup, this happens:

Screenshot 2021-02-15 at 11 11 13

I did not find an easy way to make the gray background fit the available space. I'll tinker a bit more.

**Comment by jasmussen** [comment]
Okay, got it working:

Screenshot 2021-02-15 at 11 17 14

**Comment by jasmussen** [comment]
So, in #28440, Shaun designs a "active state" for the placeholder menu item being edited. Like so:

Screenshot 2021-02-15 at 10 56 32

The inactive state has the gray background:

Screenshot 2021-02-15 at 10 56 27

But we have no such active state:

Screenshot 2021-02-15 at 11 18 35

Should we have one? 

---

Another thing. From looking at this (https://github.com/WordPress/gutenberg/pull/28861/files#diff-161a9737e181e56e07148b7eee4795f129aa821f73993a276cf1fd9c17941b70R388) there's a click handler to open the link dialog. But this handler doesn't appear to open if I set keyboard focus, and press "Enter", so we probably need a different handler that works with both mouse and keyboard:

Screenshot 2021-02-15 at 11 18 35

---

Finally, there's a conversation on verbiage. In the example here, I have a Page link, a Post link, and a custom Link. You can see the differentiation in menu item:

menu

In his mockups. Shaun showed different labels for each. George is it feasible, without a great deal of complexity, to show different placeholder text depending on the item type? So instead of "Missing URL", you'd get:

- Page Link: Select a page
- Post Link: Select a post
- Link: Add a link

@shaunandrews I trust you to correct my reading of the above ☝️

**Comment by jasmussen** [comment]
Colorized the link control separator for when the style is "alternate":

Screenshot 2021-02-15 at 11 28 31

File: packages/block-editor/src/components/url-input/index.js
Comment (georgeh) [inline packages/block-editor/src/components/url-input/index.js:105]: Found this during debugging, if you quickly toggle the `URLInput` before the fetch returns, it will attempt to call `setState()` on the unmounted component. This cancels the `suggestionsRequest` Promise when it unmounts to avoid the issue.

I can break it out to a separate PR if needed, the rest of the code doesn't trigger this anymore.

**Comment by georgeh** [comment]
@jasmussen I worked with @gwwar and we were able to get the chooser menu to trigger on keyboard navigation, add highlighting when the nav menu is not selected, and update the placeholder text. Take a look and let me know what you think.

https://user-images.githubusercontent.com/128240/108095707-28406180-704e-11eb-8d2a-981b6790a55c.mov

**Comment by georgeh** [comment]
@shaunandrews Joen's AFK for a few days, if you have time I'd appreciate your thoughts.

**Comment by shaunandrews** [comment]
This is coming together nicely!

Here's a few notes:

image

When I click the empty link block, the block itself is focused. Instead, I'd expect focus to move to the search field within the popover. Also, the link popover seems mis-aligned with the link block; I'd expect them to align to their left edges.

--

When I use my keyboard to select the empty link block, the Link UI immediately opens. This feels wrong. Instead, I'd expect to have to hit `return` to open the Link UI and then have my focus move into the popover.

--

Perhaps unrelated to this PR, but once I move my focus into the Link UI popover I have no way to close the popover. I expect that I could hit `esc` to close the popover and put my focus back on the empty link block, but it doesn't work.

--

image

The empty state is a little weird in the horizontal variation of the Navigation block, where the width of the link block is much larger than the empty state shape. Perhaps this is expected — and if you put it in a Columns block, its feels much better — I just wanted to note the strangeness.

**Comment by jasmussen** [comment]
I pushed a little fix to the border and the margin:

Screenshot 2021-02-22 at 12 06 59

The focus ring is now correctly spaced all around the placeholder, and the border is inset so it has the same size as the gray solid blob. 

These two:

> When I click the empty link block, the block itself is focused. Instead, I'd expect focus to move to the search field within the popover. Also, the link popover seems mis-aligned with the link block; I'd expect them to align to their left edges.

> When I use my keyboard to select the empty link block, the Link UI immediately opens. This feels wrong. Instead, I'd expect to have to hit return to open the Link UI and then have my focus move into the popover.

Great point. 

- Click with a mouse on the placeholder, and focus is taken to the URL field.
- Select with keyboard purely sets focus on the button. Enter or Space when there, should do the same as a click.

Sorry if I misled on this before.

> The empty state is a little weird in the horizontal variation of the Navigation block, where the width of the link block is much larger than the empty state shape. Perhaps this is expected — and if you put it in a Columns block, its feels much better — I just wanted to note the strangeness.

I'll take a look.

**Comment by jasmussen** [comment]
> The empty state is a little weird in the horizontal variation of the Navigation block, where the width of the link block is much larger than the empty state shape. Perhaps this is expected — and if you put it in a Columns block, its feels much better — I just wanted to note the strangeness.

```
.wp-block-navigation.is-vertical wp-block-navigation-link {
	display: inline-block;
}
```

that'll fix it — but I'm tempted to think we should fix this separately, because it's a general issue with the navigation block menu items:

focus size

And it does kind of open the question: what should the default behavior for that vertical version be? Happy to make a PR to add this separately, let me know!

**Comment by jasmussen** [comment]
George, this PR is in good shape! It needs a rebase, and then Shaun's point about focus addressed, then we can land it. Great work!

**Comment by georgeh** [comment]
Right now I'm reworking how the focus events happen, which should address Shaun's points. What I'm trying to do is:

Given the URL is not set:
- if you click on the block, it will pop open the URL popover
- if you select the block with the keyboard _and_ then hit return, the URL popover will open
- when the URL popover opens, the text field will get focus
- when you press escape, the URL popover closes

Once that all works I'll get another review and get this merged

**Comment by georgeh** [comment]
@jasmussen and @shaunandrews this should be ready to test

**Comment by jasmussen** [comment]
This is really nice:

nav

I can't quite put my finger on it, but there appear to be some keyboarding issues, I think it may be general to the navigation block. I'd appreciate a @shaunandrews look on this.

But specifically with mouse usage, focus remains on the button, and is not transferred to the input field. Observe the difference between the appender and the placeholder:

focus

Notice how clicking the appender sets focus in the search field, and pressing escape closes it. Clicking the page placeholder does not. What does the appender do here to make this happen, and can we copy it?

**Comment by georgeh** [comment]
When I was testing, clicking and keyboarding both triggered focus on the URL input in Firefox DE. I'll test more in Chrome and with the block variations.

**Comment by shaunandrews** [comment]
> I can't quite put my finger on it, but there appear to be some keyboarding issues, I think it may be general to the navigation block. I'd appreciate a @shaunandrews look on this.

I mentioned the keyboard focus thing earlier, and can confirm the focus still sticks with the placeholder — I'd expect it to move into the popover and focus the search input.

Related to this, once the popover does appear, its impossible to get rid of it without selecting another block. This is probably not a big deal, but as a keyboard user, I'd expect some way to get _back_ to block itself; I instinctually try hitting `esc` and this just switches to select mode. With other popovers, hitting `esc` closes the popover and puts my focus back to whatever triggered the popover.

I'd expect the placeholder state to work like a toggle. Click it one, and the popover opens and my focus moves to search box. Click it a second time and the popover should close and my focus should move back to the block.

--

Very minor, but I think there should be a `cursor: pointer` applied to the placeholder state. Right now my cursor doesn't change, so its not obvious I can click on the placeholder.

File: packages/block-library/src/navigation-link/edit.js
Comment (shaunandrews) [inline packages/block-library/src/navigation-link/edit.js:470]: I'm making this same change over in https://github.com/WordPress/gutenberg/pull/28755

I'm not sure its super relevant to this PR, so I question its inclusion here. But also, it seems there might be some discussion around wether this change is good or not.

**Comment by georgeh** [comment]
@shaunandrews can you take another look? There was a regression when I rebased earlier.

**Comment by jasmussen** [comment]
This is what I see:

now

It feels like it's perfect to me now! And it's a _vast_ improvement over what we had before, night and day. This is great.

**Review by jasmussen — APPROVED** [review]
Due to the "is-alternate" feedback from Shaun's PR, let's also remove it here (see https://github.com/WordPress/gutenberg/pull/28861#discussion_r582211556) — we can alway revisit the heuristic for when to show one or the other. So you can remove that, but otherwise, this is great.

If she has time, I'd also like one quick sanity check from @gwwar, but outside of that, this is delicious. Let's land it as soon as we can!

File: packages/block-library/src/navigation-link/edit.js
Comment (georgeh) [reply packages/block-library/src/navigation-link/edit.js:466]: Removed, per Joen

File: packages/block-library/src/navigation-link/edit.js
Comment (gwwar) [inline packages/block-library/src/navigation-link/edit.js:358]: I think this default works 👍  If we'd like CPT or custom Taxonomy labels later on, I think we end up needing new post type/taxonomy labels in WP

File: packages/block-editor/src/components/url-input/index.js
Comment (gwwar) [reply packages/block-editor/src/components/url-input/index.js:105]: Probably okay to sneak in, maybe just leave a quick note in the summary. eg "also fixes ..."

Minor: We can remove the if check using optional chaining.

```
this?.suggestionsRequest?.cancel?.()
```

See also: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining#stacking_the_optional_chaining_operator

File: packages/block-library/src/navigation-link/edit.js
Comment (gwwar) [reply packages/block-library/src/navigation-link/edit.js:358]: I think we've been leaving translator comments inline so something like the following:
```jsx
/* translators: label for missing values in navigation link block */
missingText = __( 'Add a link' );
```

**Comment by gwwar** [comment]
For folks reviewing adding a `?w=1` will ignore whitespace changes on the diff view. https://github.com/WordPress/gutenberg/pull/28861/files?w=1

**Review by gwwar — APPROVED** [review]
Great stuff @georgeh. This feels like a great improvement! I left a few minor notes, but I think this is pretty good to go. 

https://user-images.githubusercontent.com/1270189/109188311-aa641080-7747-11eb-9ed2-162e7fb53745.mp4

**Review by shaunandrews — APPROVED** [review]
Looks good! Nice work.

File: packages/block-library/src/navigation-link/edit.js
Comment (georgeh) [reply packages/block-library/src/navigation-link/edit.js:358]: I think that we could make this more extensible in the future by moving the text to `variations.js`, but there are a bunch of other bits we'd want to make extensible as well so for now I prefer this implementation.

**Comment by georgeh** [comment]
TIL you can use `?.()` to optionally call a method, thanks @gwwar 

Clear to merge?

**Comment by gwwar** [comment]
Good to go once checks pass ✅

**Pull request merged**
